Provide generic type element annotation for RepeatedField and Map types

CodingAnarchy opened the pull request:
Resolves #60 by adding the generic element types to the `::Google::Protobuf::RepeatedField` and `::Google::Protobuf::Map` output for setter methods, similarly to how it is done for `T::Hash` and `T::Array` on the getter methods.

cb-heimdall commented:
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>
<!-- CODEOWNERS & Bar Raisers -->
<!-- Power Rangers -->
<!-- Links -->

CodingAnarchy commented:
@jez Sorbet indicates that these are generic types, based on both the error it reports and [its guide to implementing the Enumerable interface](https://sorbet.org/docs/stdlib-generics#implementing-the-enumerable-interface), which indicates that effectively every `Enumerable` is a generic type. The C definition of `RepeatedField` and `Map` have [included that module for years.](https://github.com/protocolbuffers/protobuf/blob/055297bc2e66b94aad201cb62c12f9ec1b5d4b3a/ruby/ext/google/protobuf_c/repeated_field.c#L690)

> Can you share the definition of those two classes in your codebase which indicates that they've been made generic?

I'm not sure what you're asking for. We are in the process of removing some custom definitions and instead relying on the protobuf generated code for our clients, but this error is the result. None of the code that is involved here is actually _in_ our codebase. Just messages with `repeated string ...` and `map<string, string> ...` in their field definitions.

paulodiovani commented:
I tested this change by adding some fields on `testdata/example.proto`, which I suggest to be included in this PR.
Here is the diff file:

```diff
diff --git a/testdata/example.proto b/testdata/example.proto
index ae0e9bc..5ff34a3 100644
--- a/testdata/example.proto
+++ b/testdata/example.proto
@@ -6,6 +6,10 @@ package example;
 message Request {
   // some description for name field
   string name = 1;
+  // some repeated value
+  repeated string repeated_value = 2;
+  // some map
+  map<string, string> map_value = 3;
 }
 
 // some description for responsee message that is multi line and has a # in it
diff --git a/testdata/example_pb.rbi b/testdata/example_pb.rbi
index cd276fe..c35d631 100644
--- a/testdata/example_pb.rbi
+++ b/testdata/example_pb.rbi
@@ -1,5 +1,5 @@
 # Code generated by protoc-gen-rbi. DO NOT EDIT.
-# source: example.proto
+# source: testdata/example.proto
 # typed: strict
 
 # some description for request message
@@ -9,11 +9,15 @@ class Example::Request
 
   sig do
     params(
-      name: T.nilable(String)
+      name: T.nilable(String),
+      repeated_value: T.nilable(T::Array[String]),
+      map_value: T.nilable(T::Hash[String, String])
     ).void
   end
   def initialize(
-    name: ""
+    name: "",
+    repeated_value: [],
+    map_value: ::Google::Protobuf::Map.new(:string, :string)
   )
   end
 
@@ -32,6 +36,36 @@ class Example::Request
   def clear_name
   end
 
+  # some repeated value
+  sig { returns(T::Array[String]) }
+  def repeated_value
+  end
+
+  # some repeated value
+  sig { params(value: ::Google::Protobuf::RepeatedField[String]).void }
+  def repeated_value=(value)
+  end
+
+  # some repeated value
+  sig { void }
+  def clear_repeated_value
+  end
+
+  # some map
+  sig { returns(T::Hash[String, String]) }
+  def map_value
+  end
+
+  # some map
+  sig { params(value: ::Google::Protobuf::Map[String, String]).void }
+  def map_value=(value)
+  end
+
+  # some map
+  sig { void }
+  def clear_map_value
+  end
+
   sig { params(field: String).returns(T.untyped) }
   def [](field)
   end
diff --git a/testdata/example_services_pb.rbi b/testdata/example_services_pb.rbi
index 6eae4fd..a4e15a5 100644
--- a/testdata/example_services_pb.rbi
+++ b/testdata/example_services_pb.rbi
@@ -1,5 +1,5 @@
 # Code generated by protoc-gen-rbi. DO NOT EDIT.
-# source: example.proto
+# source: testdata/example.proto
 # typed: strict
 
 module Example::Greeter
```

I have also tested the changes with our custom `gem` and the fix solved the `srb tc` issues.

cb-heimdall commented:
Review Error for paulodiovani @ 2025-02-27 19:45:02 UTC
User must have write permissions to review

jez commented:
> based on both the error it reports and [its guide to implementing the Enumerable interface](https://sorbet.org/docs/stdlib-generics#implementing-the-enumerable-interface), which indicates that effectively every `Enumerable` is a generic type

The guide does mention that this is the recommended way to implement the `Enumerable` module. But the enumerable module is special, and is the only generic class or module in the Ruby standard library that is intended to be subclassed in this way, so Sorbet has special support for it. Specifically, Sorbet is fine if a class that includes `Enumerable` omits the definition of the `type_member`, so long as the file is `# typed: true` or below:

https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20%3A%3AGoogle%3A%3AProtobuf%3A%3ARepeatedField%0A%20%20include%20%3A%3AEnumerable%0A%20%20def%20each%28%26blk%29%3B%20end%0Aend%0A%0Asig%20%7B%20params%28rf%3A%20Google%3A%3AProtobuf%3A%3ARepeatedField%29.void%20%7D%0Adef%20example%28rf%29%0Aend

(Once you get up to `# typed: strict`, it becomes an error.) This is explicitly a choice to allow for cases like this, where there are otherwise-fine usages of `Enumerable` but which do not `extend T::Generic`, and thus cannot use `[]` as syntax to apply the generic to type arguments.

> Sorbet indicates that these are generic types

Sorbet does not bundle any definitions for classes defined in the `Google::Protobuf` namespace. If Sorbet is reporting that it's generic, it's due to an RBI file in your codebase.

The upstream `Google::Protobuf::{Map,RepeatedFields}`, while they include `Enumerable`, are not declared as generic: they do not have `extend T::Generic` in them nor do they declare any `type_member`s.

This means that the syntax `Google::Protobuf::RepeatedField[Integer]` is a runtime error, because the `[]` method does not exist on those classes (they do not `extend T::Generic` at runtime).

My guess is that you have an RBI file which declares them as generic.

That's fine, but not everyone that uses this gem will have those in the RBI file, because it would amount to either lying about the types (pretending that this `[]` method exists at runtime when it doesn't) or monkey patching classes provided by that gem to make this method exist at runtime. In either case, it's not necessarily a decision that can be forced on all users of this gem safely.

My suggestion would be that you build support for this behind a boolean flag, similar to how the changes in https://github.com/mdehoog/protoc-gen-rbi/pull/44 work.

> None of the code that is involved here is actually in our codebase

Your original issue report mentions a source location like `sorbet/rbi/gems/[internal].rbi:1109` which points to a piece of code that exists in your codebase.

CodingAnarchy commented:
@jez Ok, I think I understand what you are saying, but I believe that the `Enumerable` case is causing an autogenerated file to treat the repeated field as generic. This is what is generated for us at `sorbet/rbi/dsl/google/protobuf/repeated_field.rbi`:

```
# typed: true

# DO NOT EDIT MANUALLY
# This is an autogenerated file for dynamic methods in `Google::Protobuf::RepeatedField`.
# Please instead update this file by running `bin/tapioca dsl Google::Protobuf::RepeatedField`.


class Google::Protobuf::RepeatedField
  extend T::Generic

  Elem = type_member
end
```

Likewise for the map at `sorbet/rbi/dsl/google/protobuf/map.rbi`:

```
# typed: true

# DO NOT EDIT MANUALLY
# This is an autogenerated file for dynamic methods in `Google::Protobuf::Map`.
# Please instead update this file by running `bin/tapioca dsl Google::Protobuf::Map`.


class Google::Protobuf::Map
  extend T::Generic

  Key = type_member
  Value = type_member
end
```

My confusion likely stems from the fact that these are not files that we are managing - they are automatically generated by the standard `tapioca dsl` on a codebase that includes `google-protobuf` (at least at our version of 4.29.3, though the generated file has persisted since it was initially created 3 years ago without changes). In other words, we didn't make the decision for these classes to be treated as generic - _the standard Sorbet tooling did_. I would expect that this plugin would be aiming at conforming to the expectations of that tooling, which is why I aimed to make it a universal change.

> My suggestion would be that you build support for this behind a boolean flag, similar to how the changes in https://github.com/mdehoog/protoc-gen-rbi/pull/44 work.

Sure, I could work on making this change. If that's the expectation, I can start tomorrow and let you know when it is ready for another review.

jez commented:
> the standard Sorbet tooling did

Thanks for clarifying. That sounds like something to bring up with the maintainers of https://github.com/Shopify/tapioca. While we do recommend that projects use Tapioca to manage their RBI dependencies, the Sorbet team does not directly influence what tapioca does when generating RBIs—that's managed by Shopify.



> > My suggestion would be that you build support for this behind a boolean flag, similar to how the changes in #44 work.
> 
> Sure, I could work on making this change. If that's the expectation, I can start tomorrow and let you know when it is ready for another review.

Thanks!

CodingAnarchy commented:
@jez I have updated this PR to use a boolean flag, per your suggestion. I also added some test files to cover both cases with the flag and without it. Please let me know if there are any other comments or suggestions you have before we can get this merged. Appreciate it!

jez commented on ruby_types/ruby_types.go:
```diff
@@ -114,23 +114,34 @@ func rubyFieldType(field pgs.Field, mt methodType) string {
-	if mt == methodTypeSetter {
+func rubyFieldMapType(field pgs.Field, ft pgs.FieldType, mt methodType, genericContainers bool) string {
+	// A Ruby hash is not accepted at the setter
+	if mt == methodTypeSetter && !genericContainers {
```
Trying this out in the console: it looks like maybe the constructor *only* takes Array/Hash, but the getter/setter always operate on RepeatedField/Map?

In that case it suggests that the change in this PR would be to also have the `genericContainers` make the getter type use the generic container, instead of the `T::Array`/`T::Hash` types that they use (and leave the constructor alone)

CodingAnarchy replied:
I think you are right - the constructor expects a Ruby enumerator, but the other two both return the Google type. This change makes sense to me, so we should be good to merge. Thanks for improving the PR!

jez reviewed (approved):
Thank you for adding this behind a flag!

I've pushed up some changes to your PR which implement my suggestions. If they're okay with you lmk and I'll merge.

paulodiovani commented:
I tested the changes and worked as expected. Thanks. :)

Do you know when we'll get a new release version?

jez commented:
There’s a release now
